test/functional/feature_blocksxor.py

```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env python3
+# Copyright (c) 2024 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test support for XORed block data and undo files (`-blocksxor` option)."""
+import os
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.test_node import ErrorMatch
+from test_framework.util import (
+    assert_equal,
+    assert_greater_than,
+    read_xor_key,
+    util_xor,
+)
+from test_framework.wallet import MiniWallet
+
+
+class BlocksXORTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.extra_args = [[
+            '-blocksxor=1',
+            '-fastprune=1',             # use smaller block files
+            '-datacarriersize=100000',  # needed to pad transaction with MiniWallet
+        ]]
+
+    def run_test(self):
+        self.log.info("Mine some blocks, to create multiple blk*.dat/rev*.dat files")
+        node = self.nodes[0]
+        wallet = MiniWallet(node)
+        for _ in range(5):
+            wallet.send_self_transfer(from_node=node, target_weight=80000)
+            self.generate(wallet, 1)
+
+        block_files = list(node.blocks_path.glob('blk[0-9][0-9][0-9][0-9][0-9].dat'))
+        undo_files  = list(node.blocks_path.glob('rev[0-9][0-9][0-9][0-9][0-9].dat'))
+        assert_equal(len(block_files), len(undo_files))
+        assert_greater_than(len(block_files), 1)  # we want at least one full block file
+
+        self.log.info("Shut down node and un-XOR block/undo files manually")
+        self.stop_node(0)
+        xor_key = read_xor_key(node=node)
+        for data_file in sorted(block_files + undo_files):
+            self.log.debug(f"Rewriting file {data_file}...")
+            with open(data_file, 'rb+') as f:
+                xored_data = f.read()
+                f.seek(0)
+                f.write(util_xor(xored_data, xor_key, offset=0))
+
+        self.log.info("Check that restarting with 'blocksxor=0' fails if XOR key is present")
+        node.assert_start_raises_init_error(['-blocksxor=0'],
+            'The blocksdir XOR-key can not be disabled when a random key was already stored!',
+            match=ErrorMatch.PARTIAL_REGEX)
```

Comment:
This fails for me locally, because somehow I have 0000000000000000 as my xor key... but it fails consistently. Unsure if that's supposed to be possible?

Reply:
I added a `if (opts.use_xor) assert(xor_key != decltype(xor_key){});` which hit, so not an old bitcoind. Attaching combined log.
```
 node0 2024-08-15T16:40:41.093882Z [init] [node/blockstorage.cpp:1184] [InitBlocksdirXorKey] Using obfuscation key for blocksdir *.dat files (/tmp/bitcoin_func_test__c38k83u/node0/regtest/blocks): '0000000000000000' 
```
[logs_30657.txt](https://github.com/user-attachments/files/16627687/logs_30657.txt)

Reply:
[logs_30657_2.txt](https://github.com/user-attachments/files/16627939/logs_30657_2.txt)

Ah interesting, `fs::is_empty(opts.blocks_dir)` is 0

Reply:
It's jammy with clang 17, indeed it works with a non-`/tmp` folder.

Reply:
```
Filesystem      Size  Used Avail Use% Mounted on
/dev/nvme0n1p2  468G  110G  335G  25% /
```
It worked with gcc. `./configure` didn't like libc++-17. After reboot I can't get it to fail anymore (also I don't mean to hold up this PR)

Reply:
I tried this on my apple machine and am able to make this test fail if I add a sleep right before the line checking `fs::is_empty`. No sign of DS_Store. It kind of seems more like blk0000.dat, rev0000.dat, and index/ are showing up early... I don't think I know init well enough to know exactly what order everything should happen in.

Reply:
Found this here because I saw #30833 and it got me to look into `blocksxor` for the first time. I see the same failures @glozow describes (also on macos) and I can confirm the findings here about the use of the temp folders because it does not happen when the test uses a clean chain, e.g. using a clean chain like this made the test robust for me: https://github.com/fjahr/bitcoin/commit/ef00b8eab48cf7ebcc8a0539f237175b8f5a0373  I didn't find anything new beyond that so far but I thought it should be noted that it's not on @glozow 's machine alone.

Reply:
> I guess fs::is_empty(opts.blocks_dir) could be replaced by not fs::exists("blk0000.dat")?

I think that alone wouldn't work if the user upgrades a pruned node to v28?

Anyway, I was going to revisit this today but I simply don't see the failures anymore currently 🤷‍♂️  I will get back to it if I see them occur again.